pkg/messagebatcher/batchwriter.go

```diff
@@ -0,0 +1,396 @@
+				return
+			}
+			bw.addBatch(msg, more)
+		}
+	}
+}
+
+func (bw *batchWriter) closeCurBatch() {
+	bw.mu.Lock()
+	defer bw.mu.Unlock()
+	if bw.curBatch != nil {
+		bw.curBatch.ready <- struct{}{}
+		close(bw.curBatch.ready)
+	}
+}
+
+func (bw *batchWriter) addBatch(msg *Message, more bool) {
+	bw.mu.Lock()
+	defer bw.mu.Unlock()
+	if bw.curBatch == nil {
+		bw.curBatch = bw.newBatch(bw.cfg.msgInterval, bw.cfg.sizeLimit)
+	}
+	bw.extendTimeout()
+	bw.curBatch.Add(msg)
+	if bw.curBatch.Full() {
+		bw.curBatch.Flush()
+		bw.curBatch = nil
+	}
+}
+
+func (bw *batchWriter) newBatch(msgInterval time.Duration, limit uint64) *batch {
+	batch := &batch{
+		msgs:      make([]*Message, 0),
+		sizeLimit: limit,
+		timer:     time.NewTimer(msgInterval),
+		ready:     make(chan struct{}),
+	}
+	go bw.awaitBatch(batch)
+	return batch
+}
+
+func (bw *batchWriter) awaitBatch(b *batch) {
+	select {
+	case <-b.timer.C:
+	case <-b.ready:
+	}
+	bw.mu.Lock()
+	if bw.curBatch == b {
+		bw.curBatch = nil
+	}
+	bw.mu.Unlock()
+	bw.manager.assembleQueue <- b
+	b.timer.Stop()
+}
+
+func (bw *batchWriter) extendTimeout() {
+	atomic.SwapUint32(&bw.timeoutTimes, 0)
+}
+
+func (bw *batchWriter) Put(msg *Message) error {
+	if !bw.IsReady() {
+		return errors.New("writer hasn't started yet")
+	}
+	select {
+	case bw.msgBuffer <- msg:
+		return nil
+	default:
+		return errors.New("the msg buffer of writer is full")
+	}
+}
+
+func (bw *batchWriter) AddTimeoutTimes() {
+	atomic.AddUint32(&bw.timeoutTimes, 1)
+}
+
+func (bw *batchWriter) Expired() bool {
+	return atomic.LoadUint32(&bw.timeoutTimes) >= bw.cfg.expiredThreshold
+}
+
+func (bw *batchWriter) Close() {
+	bw.TurnOff()
+	close(bw.msgBuffer)
+}
+
+type batch struct {
+	msgs      []*Message
+	sizeLimit uint64
+	timer     *time.Timer
+	ready     chan struct{}
+}
+
+func (b *batch) Size() int {
+	return len(b.msgs)
+}
+
+func (b *batch) Add(msg *Message) {
+	b.msgs = append(b.msgs, msg)
+}
+
+func (b *batch) Full() bool {
+	return uint64(len(b.msgs)) >= b.sizeLimit
+}
+
+func (b *batch) Flush() {
+	b.ready <- struct{}{}
+	close(b.ready)
+}
+
+// Message is the message to be batching
+type Message struct {
+	ChainID uint32
+	Data    proto.Message
+	Target  *peer.AddrInfo       // target of broadcast msg is nil
+	id      *batchID             // cache for Message.BatchID()
+	msgType iotexrpc.MessageType // generated by MessageType()
+}
+
+func (msg *Message) batchID() (id batchID, err error) {
+	if msg.id != nil {
+		id = *msg.id
+		return
+	}
+	buf := new(bytes.Buffer)
+	if err = binary.Write(buf, binary.LittleEndian, msg.messageType()); err != nil {
+		return
+	}
+	if err = binary.Write(buf, binary.LittleEndian, msg.ChainID); err != nil {
+		return
+	}
+	var idInBytes []byte
+	if msg.Target != nil {
+		idInBytes, err = msg.Target.ID.Marshal()
+		if err != nil {
+			return
+		}
+	}
+	// non-cryptographic fast hash algorithm xxhash is used for generating batchID
+	h := xxhash.Sum64(append(buf.Bytes(), idInBytes...))
```

Comment:
recommend to use siphash, which is fast hash algo developed by renowned security researcher, and used in many open source projects
1.  https://en.wikipedia.org/wiki/SipHash
2. github.com/dchest/siphash

Reply:
https://github.com/Cyan4973/xxHash/wiki/Performance-comparison